lib/virtio/virtqueue.c

```diff
@@ -697,3 +733,25 @@ static int virtqueue_navail(struct virtqueue *vq)
 
 	return navail;
 }
+
+static void *virtqueue_get_next_avail_buffer(struct virtqueue *vq, uint16_t idx,
+					     uint16_t *next_idx, uint32_t *next_len)
+{
+	void *buffer;
+	uint16_t next;
+
+	VRING_INVALIDATE(vq->vq_ring.desc[idx], sizeof(struct vring_desc));
+	if (!(vq->vq_ring.desc[idx].flags & VRING_DESC_F_NEXT))
+		return NULL;
+
+	next = vq->vq_ring.desc[idx].next;
+	if (next_idx)
+		*next_idx = next;
+
+	VRING_INVALIDATE(vq->vq_ring.desc[next], sizeof(struct vring_desc));
+	buffer = virtqueue_phys_to_virt(vq, vq->vq_ring.desc[next].addr);
+	if (next_len)
```

Comment:
Instead, I would add a test on next_len to return an error if it is null. Having a buffer without its length seems dangerous to me.

Reply:
I add a VQ_ASSERT to this condition